### ISSUE TITLE
Weird issue (U2+) Cartridge on C64 Silver label.

### ISSUE THREAD

**Issue by 0wwafa** [GideonZ/1541ultimate#439]
I have tried a few cartridges on the U2+ and they all seem to work on the breadbin.
Today I found out some cartidges I made, work on the breadbin but not on the silver label.
It's not about ROM differences between the two.

They are both magic desk type.

I have no idea how to debug this because in the emulator(s) they both work,

[archon.zip](https://github.com/user-attachments/files/17831797/archon.zip)

Attached are the two cartridges.
Any help apreciated (if it's my fault) if it's not.. then it's on you :D

**Comment by radius75** [comment]
It looks a bit like #335

**Comment by UffeJakobsen** [comment]
Have you adjusted PHI1/PHI2 settings for that Silver Label ?

**Comment by radius75** [comment]
This is probably not a problem with PHI, I have the same problem with Archon as OP.

**Comment by 0wwafa** [comment]
> Have you tried adjusted PHI1/PHI2 settings for that Silver Label ?

Yep. I tried every possible setting of PHI1/PHI2.
Probably it's because my cart loader is the fastest possible.. but I don't yet know if it's the cart loader that causes the problem or the U2+.

**Comment by 0wwafa** [comment]
> It looks a bit like #335 on my U2+

Hmm.. I did not test it with 3.9... I am on 3.11 now.

**Comment by radius75** [comment]
Archon 64KB crt doesn't work on fw3.9 for me either. So it probably never worked.

**Comment by mrdudz** [comment]
The first thing to do here would be replacing the archon payload by something like a checksummed self checking binary (then you can see if the cartridge startup is actually broken) - like this its a pretty useless test case

**Comment by radius75** [comment]
```Creating CRT type from info: WorkinginU2onSilverLabel.crt
Object level 1 returned 1.
Action set was: Run Cart
Cartridge Load.. WorkinginU2onSilverLabel.crt
CRT Hardware type: 19 Header OK. Now reading chip packets starting from 000040.
Reading chip data bank 0 ($8000) with size $2000 to 0x00F00000.
Reading chip data bank 1 ($8000) with size $2000 to 0x00F04000.
Reading chip data bank 2 ($8000) with size $2000 to 0x00F08000.
Reading chip data bank 3 ($8000) with size $2000 to 0x00F0C000.
Reading chip data bank 4 ($8000) with size $2000 to 0x00F10000.
Reading chip data bank 5 ($8000) with size $2000 to 0x00F14000.
Reading chip data bank 6 ($8000) with size $2000 to 0x00F18000.
Reading chip data bank 7 ($8000) with size $2000 to 0x00F1C000.
Reading chip data bank 8 ($8000) with size $2000 to 0x00F20000.
Reading chip data bank 9 ($8000) with size $2000 to 0x00F24000.
Reading chip data bank 10 ($8000) with size $2000 to 0x00F28000.
Reading chip data bank 11 ($8000) with size $2000 to 0x00F2C000.
Reading chip data bank 12 ($8000) with size $2000 to 0x00F30000.
Reading chip data bank 13 ($8000) with size $2000 to 0x00F34000.
Reading chip data bank 14 ($8000) with size $2000 to 0x00F38000.
Reading chip data bank 15 ($8000) with size $2000 to 0x00F3C000.
Reading chip data bank 16 ($8000) with size $2000 to 0x00F40000.
Reading chip data bank 17 ($8000) with size $2000 to 0x00F44000.
Reading chip data bank 18 ($8000) with size $2000 to 0x00F48000.
Reading chip data bank 19 ($8000) with size $2000 to 0x00F4C000.
Reading chip data bank 20 ($8000) with size $2000 to 0x00F50000.
Reading chip data bank 21 ($8000) with size $2000 to 0x00F54000.
Reading chip data bank 22 ($8000) with size $2000 to 0x00F58000.
Reading chip data bank 23 ($8000) with size $2000 to 0x00F5C000.
Reading chip data bank 24 ($8000) with size $2000 to 0x00F60000.
Reading chip data bank 25 ($8000) with size $2000 to 0x00F64000.
Reading chip data bank 26 ($8000) with size $2000 to 0x00F68000.
Reading chip data bank 27 ($8000) with size $2000 to 0x00F6C000.
Reading chip data bank 28 ($8000) with size $2000 to 0x00F70000.
Reading chip data bank 29 ($8000) with size $2000 to 0x00F74000.
Reading chip data bank 30 ($8000) with size $2000 to 0x00F78000.
Reading chip data bank 31 ($8000) with size $2000 to 0x00F7C000.
Total ROM size read: 040000 bytes.
*28*
About to execute a command in subsys C64 Machine (00688388)
Cartridge definition:
Name: Magic Desk, Domark, HES Australia
Type: 40
Custom: 00000000 (Length: 0)
Required: 0000
Prohibited: 0001
Sampler found in FPGA... IO map: disabled.
Writing 85 to timing register. 5/8/0
Cartridge registers:
04040000: 04 00 02 17 00 28 01 00 00 04 00 00 00 00 00 00 .....(..........
Begin of cart init: Type: 28. REU: 00. REU_SZ: 04, UCI: 00 (DF18), Mode: 04, Sampler: 00
End of cart init: Type: 28. REU: 00. REU_SZ: 04, UCI: 00 (DF18), Mode: 04, Sampler: 00
Draw. No window to draw on.
Ultimax set.. Now reading registers..
Mode=1
Frozen on Bad line. Raster = 82. VIC Irq Enable: F1. Vic IRQ: F1
Original d012/11 content = 57 3B.
CIA1 registers: FE FF FF 00 FF FF FF FF 00 00 00 91 00 
Init IO.
CIA DDR: 00 FF Mode: 02
CIA DDR: 00 FF
State Temp reloaded. # of children = 2
Number of items: 11
@Action Category selected: Developer
Object level 2 returned 1.
Object level 1 returned 1.
Action set was: Save Debug Log
```

```Creating CRT type from info: NotWorkinginU2onSilverLabel.crt
Object level 1 returned 1.
Action set was: Run Cart
Cartridge Load.. NotWorkinginU2onSilverLabel.crt
CRT Hardware type: 19 Header OK. Now reading chip packets starting from 000040.
Reading chip data bank 0 ($8000) with size $2000 to 0x00F00000.
Reading chip data bank 1 ($8000) with size $2000 to 0x00F04000.
Reading chip data bank 2 ($8000) with size $2000 to 0x00F08000.
Reading chip data bank 3 ($8000) with size $2000 to 0x00F0C000.
Reading chip data bank 4 ($8000) with size $2000 to 0x00F10000.
Reading chip data bank 5 ($8000) with size $2000 to 0x00F14000.
Reading chip data bank 6 ($8000) with size $2000 to 0x00F18000.
Reading chip data bank 7 ($8000) with size $2000 to 0x00F1C000.
Total ROM size read: 010000 bytes.
*28*
About to execute a command in subsys C64 Machine (00688388)
Cartridge definition:
Name: Magic Desk, Domark, HES Australia
Type: 40
Custom: 00000000 (Length: 0)
Required: 0000
Prohibited: 0001
Sampler found in FPGA... IO map: disabled.
Writing 85 to timing register. 5/8/0
Cartridge registers:
04040000: 04 00 02 17 00 28 01 00 00 04 00 00 00 00 00 00 .....(..........
Begin of cart init: Type: 28. REU: 00. REU_SZ: 04, UCI: 00 (DF18), Mode: 04, Sampler: 00
End of cart init: Type: 28. REU: 00. REU_SZ: 04, UCI: 00 (DF18), Mode: 04, Sampler: 00
Draw. No window to draw on.
@Hard stop!!!
CIA1 registers: FF FF 00 00 FF FF FF FF 00 00 00 01 00 
Init IO.
CIA DDR: 00 FF Mode: 02
CIA DDR: 00 FF
State Temp reloaded. # of children = 3
Number of items: 11
Action Category selected: Developer
Object level 2 returned 1.
Object level 1 returned 1.
Action set was: Save Debug Log```

**Comment by radius75** [comment]
I don't know. Only @GideonZ  can read Ultimate logs.
Is your Archon available for download from any website? Was this game officially released on a cartridge? Is this your own version?

**Comment by 0wwafa** [comment]
> I don't know. Only @GideonZ can read Ultimate logs. Is your Archon available for download from any website? Was this game officially released on a cartridge? Is this your own version?

It's my own version. But it's not just archon. I used the same loader for other games. They all don't work in the silver but work in the breadbin.
So it's the loader. But I don't know why.
Aside from a more unrolled loop (so faster loading) there are no difference that I can see with the working ones.

**Comment by 0wwafa** [comment]
I am very curious if the loader is triggering a bug in the U2 or some other reason (and if so, I want to know what is the reason).
Because I can't understand it.

**Comment by 0wwafa** [comment]
What I do know is that it's not related to the different Silver ROM. Otherwise the cart would fail in VICE using the Silver ROMS.
(Also the code is quite simple and does not do anything crazy).
I suspect some timing problem between U2 and PLA but I don't know enough to diagnose that.

**Comment by mrdudz** [comment]
It is most certainly something like an uninitialized variable somewhere, and unrelated to the 1541U or the Silver Label C64 :) As said, a self checking payload will show it.

**Comment by radius75** [comment]
@0wwafa 
Your Archon doesn't work on my C64C, short motherboard.

**Comment by 0wwafa** [comment]
> @0wwafa Your Archon doesn't work on my C64C, short motherboard. This is visible in the log I pasted `@Hard stop!!!`

That's even weirder.
I have no clue about WHY.

**Comment by 0wwafa** [comment]
> It is most certainly something like an uninitialized variable somewhere, and unrelated to the 1541U or the Silver Label C64 :) As said, a self checking payload will show it.

The CRTs work on VICE and other emulator (DENISE).

They also work on breadbin.

If it was some uninitialized variable they would give random problems but they don't.

**Comment by mrdudz** [comment]
> If it was some uninitialized variable they would give random problems but they don't.

No. It can still be that and just "work" in most cases, by chance. You'd be surprised how many original VIC20 programs exist that rely on uninitialized variables - and yet they work (almost) everywhere :)

> About a self checking program, if you have such program I will be happy to try it.

cgen from https://csdb.dk/release/?id=226382

**Comment by 0wwafa** [comment]
> > If it was some uninitialized variable they would give random problems but they don't.
> 
> No. It can still be that and just "work" in most cases, by chance. You'd be surprised how many original VIC20 programs exist that rely on uninitialized variables - and yet they work (almost) everywhere :)
> 
> > About a self checking program, if you have such program I will be happy to try it.
> 
> cgen from https://csdb.dk/release/?id=226382

I understand that, but my code has no uninitialized variables.

@radius75 it won't work from pendrive (in most cases) you must say yes when the u2 asks if you want to copy the CRT in the flash.

Only then it will work (on bradbin at least).

This is another reason that makes me suspect the U2 misbehaving with timings.

**Comment by 0wwafa** [comment]
Also, for reference I used [this firmware](https://github.com/Zibri/1541-Ultimate-Firmware/releases/download/z7a410e3/u2plus_fw_3.11a_7a410e3Z.7z)

**Comment by mrdudz** [comment]
Perhaps you should first install the last official firmware then...

**Comment by radius75** [comment]
> Also, for reference I used [this firmware](https://github.com/Zibri/1541-Ultimate-Firmware/releases/download/z7a410e3/u2plus_fw_3.11a_7a410e3Z.7z)

Last OFW for U2+ is 3.11
Common testing only makes sense if everyone is using the same FW version. Otherwise it's a waste of time.

**Comment by radius75** [comment]
Do this:
Extract .bin from your .crt
In a hex editor, duplicate this code 4x to make it 256KB
Convert in cartconv to -t md

For me, the .crt prepared this way works on U2+ with OFW3.11

**Comment by radius75** [comment]
This is what the log looks like when I run the .crt (Yours .bin multiplied to 256KB)

```Creating CRT type from info: arch4x.crt
Object level 1 returned 1.
Action set was: Run Cart
Cartridge Load.. arch4x.crt
CRT Hardware type: 19 Header OK. Now reading chip packets starting from 000040.
Reading chip data bank 0 ($8000) with size $2000 to 0x00F00000.
Reading chip data bank 1 ($8000) with size $2000 to 0x00F04000.
Reading chip data bank 2 ($8000) with size $2000 to 0x00F08000.
Reading chip data bank 3 ($8000) with size $2000 to 0x00F0C000.
Reading chip data bank 4 ($8000) with size $2000 to 0x00F10000.
Reading chip data bank 5 ($8000) with size $2000 to 0x00F14000.
Reading chip data bank 6 ($8000) with size $2000 to 0x00F18000.
Reading chip data bank 7 ($8000) with size $2000 to 0x00F1C000.
Reading chip data bank 8 ($8000) with size $2000 to 0x00F20000.
Reading chip data bank 9 ($8000) with size $2000 to 0x00F24000.
Reading chip data bank 10 ($8000) with size $2000 to 0x00F28000.
Reading chip data bank 11 ($8000) with size $2000 to 0x00F2C000.
Reading chip data bank 12 ($8000) with size $2000 to 0x00F30000.
Reading chip data bank 13 ($8000) with size $2000 to 0x00F34000.
Reading chip data bank 14 ($8000) with size $2000 to 0x00F38000.
Reading chip data bank 15 ($8000) with size $2000 to 0x00F3C000.
Reading chip data bank 16 ($8000) with size $2000 to 0x00F40000.
Reading chip data bank 17 ($8000) with size $2000 to 0x00F44000.
Reading chip data bank 18 ($8000) with size $2000 to 0x00F48000.
Reading chip data bank 19 ($8000) with size $2000 to 0x00F4C000.
Reading chip data bank 20 ($8000) with size $2000 to 0x00F50000.
Reading chip data bank 21 ($8000) with size $2000 to 0x00F54000.
Reading chip data bank 22 ($8000) with size $2000 to 0x00F58000.
Reading chip data bank 23 ($8000) with size $2000 to 0x00F5C000.
Reading chip data bank 24 ($8000) with size $2000 to 0x00F60000.
Reading chip data bank 25 ($8000) with size $2000 to 0x00F64000.
Reading chip data bank 26 ($8000) with size $2000 to 0x00F68000.
Reading chip data bank 27 ($8000) with size $2000 to 0x00F6C000.
Reading chip data bank 28 ($8000) with size $2000 to 0x00F70000.
Reading chip data bank 29 ($8000) with size $2000 to 0x00F74000.
Reading chip data bank 30 ($8000) with size $2000 to 0x00F78000.
Reading chip data bank 31 ($8000) with size $2000 to 0x00F7C000.
Total ROM size read: 040000 bytes.
*28*
About to execute a command in subsys C64 Machine (00688388)
Cartridge definition:
Name: Magic Desk, Domark, HES Australia
Type: 40
Custom: 00000000 (Length: 0)
Required: 0000
Prohibited: 0001
Sampler found in FPGA... IO map: disabled.
Writing 85 to timing register. 5/8/0
Cartridge registers:
04040000: 04 00 02 17 00 28 01 00 00 04 00 00 00 00 00 00 .....(..........
Begin of cart init: Type: 28. REU: 00. REU_SZ: 04, UCI: 00 (DF18), Mode: 04, Sampler: 00
End of cart init: Type: 28. REU: 00. REU_SZ: 04, UCI: 00 (DF18), Mode: 04, Sampler: 00
Draw. No window to draw on.
Ultimax set.. Now reading registers..
Mode=1
Frozen on Bad line. Raster = 97. VIC Irq Enable: F1. Vic IRQ: 7E
Original d012/11 content = 46 1B.
CIA1 registers: 7F FF FF 00 FD 38 FF FF 00 00 00 01 00 
Init IO.
CIA DDR: 00 FF Mode: 02
CIA DDR: 00 FF
State Temp reloaded. # of children = 1
Number of items: 11
Action Category selected: Developer
Object level 2 returned 1.
Object level 1 returned 1.
Action set was: Save Debug Log
```

**Comment by 0wwafa** [comment]
@radius75 is not a problem of the game: NO game works with this loader.

> Last OFW for U2+ is 3.11
> Common testing only makes sense if everyone is using the same FW version. Otherwise it's a waste of time.

the firmware I linked is the latest commit from Gideon. It is indeed 3.11. I just compiled it.


> Convert in cartconv to -t md

that would create a different loader!

I repeat: my loader is faster than most. I think U2 can't handle the speed.

The problem is not related to the specific game.

**Comment by 0wwafa** [comment]
> Perhaps you should first install the last official firmware then...

It's the same. The one I linked is just more recent. But I tried also with the oficial "older" version.

**Comment by radius75** [comment]
> They are both magic desk type.

-t md converts to Magic Desk type, sorry but I completely misunderstood the problem then.

**Comment by GideonZ** [comment]
So let's go over the things that were tried. When the binary file was extended and then converted back to .crt, it actually did work. Is this correct? If that's the case, the program incorrectly programs the bank register, such that it exceeds the installed ROM size.

**Comment by mrdudz** [comment]
> the firmware I linked is the latest commit from Gideon. It is indeed 3.11. I just compiled it.

It is still not the same and even if you compiled the exact same source that was used for the official release (which you did not) the result could be a different binary, and even behave different regarding timing, due to how FPGA synthesis works (involving a PRNG). That's why it is important to use the exact same binaries.

> that would create a different loader!

absolutely not. cartconv converts into .crt format. no more no less. it has nothing to do with any "loader".

> I think that my loader is so fast that also the flash loading does not work correctly.

no way that is the problem. at all. if the implementation couldn't handle a read on every consecutive cycle, then nothing would work.

> When the binary file was extended and then converted back to .crt, it actually did work. Is this correct? If that's the case, the program incorrectly programs the bank register, such that it exceeds the installed ROM size.

good idea, but this seems to be correct:
```
MAGICDESK: Reg: 01 (Bank: 1 of 8, enabled)
MAGICDESK: Reg: 02 (Bank: 2 of 8, enabled)
MAGICDESK: Reg: 03 (Bank: 3 of 8, enabled)
MAGICDESK: Reg: 04 (Bank: 4 of 8, enabled)
MAGICDESK: Reg: 05 (Bank: 5 of 8, enabled)
MAGICDESK: Reg: 00 (Bank: 0 of 8, enabled)
MAGICDESK: Reg: 80 (Bank: 0 of 8, disabled)
```

**Comment by radius75** [comment]
> Is this correct?

I extracted the .bin like this, duplicated it to 256KB and re-converted it to Magic Desk type.

```cartconv.exe -i NotWorkinginU2onSilverLabel.crt -t bin -o arch.bin
Input file: NotWorkinginU2onSilverLabel.crt
Output file: arch.bin
Conversion from C64 Magic Desk .crt to binary format successful.
```

```cartconv.exe -i arch4x.bin -t md -o arch4x.crt
Input file: arch4x.bin
Output file: arch4x.crt
Conversion from binary format to C64 Magic Desk .crt successful.
```


Attached is the output .crt
[arch4x.zip](https://github.com/user-attachments/files/18048541/arch4x.zip)


> If that's the case, the program incorrectly programs the bank register, such that it exceeds the installed ROM size.

This would mean that Vice allows you to run such a .crt and U2+ does not.
OP also claims that this .crt works fine for him on U2+ and Breadbin. Weird.

**Comment by mrdudz** [comment]
> This would mean that Vice allows you to run such a .crt and U2+ does not.

VICE adds mirrors depending on the size of the loaded ROM - so yes, it would work IF this would be the problem :)

I can't say if this behaviour is correct - i have not seen a real Magic Desk cartridge

**Comment by GideonZ** [comment]
Whether the behavior is correct depends on the ROM that is installed on the cartridge. If one or more of the address lines end up on a 'no connect', mirrors will appear automatically. When the ROM is larger than the space that is used, it will return FF on the higher banks.

The only correct way is that the cartridge code itself does not select banks higher than it uses. It is wrong to assume that mirrors will exist.

**Comment by 0wwafa** [comment]
What I know is only that this loader is faster than others.
The game inside the CRT does not change the behavior.
They all work or not work.
And they all work in any emulator.
As I said I don't know what's wrong.
But it's not "the game".
I made Ghosts 'n goblis, archon, pharaoh's curse and giana sisters (both 64k both 128k including  the intro).

They all work in all emulators.
I can provide them all but the loader is exactly the same as in the archon cartridge I attached.
If you check the loader code with VICE machine code monitor is quite simple.
It just has a big unrolled loop to move data from the cartridge banks to memory.
And that's what makes it faster. That's all.
(Optimizing a bit the assembly code I could make it even a tad faster)

**Comment by GideonZ** [comment]
Your loader being fast or not is not relevant.

What is relevant is what you write into the IO registers of the cartridge and what behavior you expect. Nothing more, nothing less.

I am not saying that the magic desk emulation is perfect. But what I am trying to get above the table here is if it is a mirroring issue. You seem to evade the question, because I haven't seen a clear answer to it.

**Comment by radius75** [comment]
@0wwafa 
Converting .bin and duplicating x4 data should not have any effect on your loader which is saved in .bin

If the loader refers to mirrored locations then the duplicated x4 .bin created 3 additional "physical" locations, instead of mirrored ones. That's how I understand it.

Maybe just double the . bin and do it once and your loader for the 128KB . crt will work fine
I haven't checked that.

If I understand correctly the mirrored location depends on the size of the RAM chips used (real hardware)
It is implemented differently in Vice and differently in Ultimate.

I managed to run quite a few .bin files ripped from niche utility cartridges on Ultimate by duplicating their content.
It probably depends on what size RAM chips they were obtained from.
Only now I understood why they worked immediately in Vice and on Ultimate only after doubling. Their loader was prepared for a specific size of memory chip or they were saved in it in that memory chip multiple times.

I hope I'm not too far wrong in my understanding.

**Comment by 0wwafa** [comment]
@GideonZ if it helps this is the heart of the loader:
It does not do anything particular: it increments a bunch of addresses of the unrolled loop, the increments the bank until it reaches 16 or 32 (it's modified by the generator)
I even put the STA in $1 quite far from each other so that they don't happen too soon from each other (just in case).

```
do_next:                                ; CODE XREF: RAM:81B8↑j
                LDA     #$37
                STA     1
                INC     $FB
                LDY     $FB
;bank3=*+1-cold_start+9
bank2=*+1-boot_100+$100
                CPY     #$10
                BEQ     end_reloc

                LDA #<naddr
                STA jaddr

                STY     $DE00
                LDA     #$33
                STA     1
                LDX     #7


update_addresses:                       ; CODE XREF: RAM:81E0↓j
                LDA     $FC
                STA     $100,X
                INC     $FC
                TXA
                CLC
                ADC     #6
                TAX
                CMP     #$C7
                BNE     update_addresses
                JMP     $100   ;  here it goes back to the main loop.
```

**Comment by markusC64** [comment]
I have wset a "break store de00 dfff" in VICE and saw:

.C:8012  A9 08       LDA #$08
.C:8014  8D 00 DE    STA $DE00

Later you can find

.C:8192  A9 08       LDA #$08
.C:8194  8D 00 DE    STA $DE00
.C:8197  60          RTS

Both are clearly wrong since your cartridge only has 7 banks.

**Comment by 0wwafa** [comment]
Sincerely I don't remember why I did that.
I will recheck it. but anyway #$08 = 00 if the cartridge has 7 banks.

**Comment by markusC64** [comment]
>  but anyway #$08 = 00 if the cartridge has 7 banks.

Not on the ultimate as explained above by others.

And your cartridge does not work on the ultimate 64, too. But if I patch both occurencies of "A9 08 8D 00 DE" to "A9 00 8D 00 DE" then the gams starts on the ultimate 64.

**Comment by 0wwafa** [comment]
> > but anyway #$08 = 00 if the cartridge has 7 banks.
> 
> Not on the ultimate as explained above by others.
> 
> And your cartridge does not work on the ultimate 64, too. But if I patch both occurencies of "A9 08 8D 00 DE" to "A9 00 8D 00 DE" then the gams starts on the ultimate 64.

There is ONLY one occurrence (at $8012). The second one is overwritten by the code itself.
Which is a typo. (But harmless because bank #8 = bank #0).
But if on u2 that is not implemented than it explains it.

```
.C:8012  A9 00       LDA #$00
.C:8014  8D 00 DE    STA $DE00
```

**Comment by 0wwafa** [comment]
I rechecked the code. That #$08 is not a typo.. is there because if the cartridge is 64k it points to bank #0 but if the cartridge is 128K then it points to bank #8.

So: in the end the reason (thanks for that) is that the u2 does not support the mirroring (which happens on a real cartridge).

**Comment by radius75** [comment]
One thing is still unanswered. So how does this 64KB .crt work on your Breadbin with U2+? 🤔

**Comment by 0wwafa** [comment]
> One thing is still unanswered. So how does this 64KB .crt work on your Breadbin with U2+? 🤔

that is a mystery... there is a slight chance I used a previous version which has 00 (the loader was only meant for 64k cartridges) then I made a second version that supports 64k and 128K.

On real cartridge if a bank is not there the address becomes #00 and it works. On U2 didn't and I did not notice.
I assumed the U2 was working as a normal cartridge.
Thanks to you I understood. But I still think it's U2 fault because on a real magicdesk cartridge bank #8 = bank #0.

**Comment by 0wwafa** [comment]
Thanks everyone for helping me understand.
It was driving me crazy.

**Comment by GideonZ** [comment]
> So: in the end the reason (thanks for that) is that the u2 does not support the mirroring (which happens on a real cartridge).

This cannot be said, mr Zambrini. What if the cartridge has a 128K rom, but only the first 64K area is programmed with your ROM data?

The mirroring in real hardware only occurs when some address lines do not exist on the ROM.

**Comment by 0wwafa** [comment]
@GideonZ 
the error seems to be here:
```
    bank &= max_bank;

    uint32_t offset = uint32_t(bank) * bank_multiplier;
```
https://github.com/GideonZ/1541ultimate/blob/7a410e38b607021296b462ecfd9d2af66ee437ef/software/io/c64/c64_crt.cc#L274

or around there.

**Comment by 0wwafa** [comment]
@GideonZ 
looks correct but max_bank seems to be always FF instead max_bank should be taken by the CRT header.
that's all.

**Comment by GideonZ** [comment]
@0wwafa It won't fix a thing. The problem you observe is caused by the way the hardware works. MagicDesk is 512K or even 1MB. When you simply write the wrong bank value, it will select the wrong memory area, as simple as that. The only way to fix this is to create mirrors in software by applying memcpy.

**Comment by 0wwafa** [comment]
> @0wwafa It won't fix a thing. The problem you observe is caused by the way the hardware works. MagicDesk is 512K or even 1MB. When you simply write the wrong bank value, it will select the wrong memory area, as simple as that. The only way to fix this is to create mirrors in software by applying memcpy.

Wrong: Magic Desk can have ANY size, 64K,128K, 256K, 512K and 1MB.

P.S.
Why do you create mirrors by copying memory when you can use the max_bank when you get the block in `read_chip_packet` ?
It would be very easy just to set "max_bank" which at the moment is instead useless because it's always 255.

But it's your software and you do as you like, I'll add it to my own patches.
An emulator should emulate the original hardware behaviour.

> Magic Desk cartridges, in their original form, could hold up to 128Kb of code, or up to 16 banks of 8Kb each.

Where UP TO means they could also be 64K.
source: https://github.com/msolajic/c64-magic-desk-512k

Here is another:
https://github.com/RetroNynjah/Magic-Desk-128

and there are more and more:
https://www.pcbway.com/project/shareproject/C64_64KB_MAGIC_DESK_COMPATIBLE_SINGLE_FILE_AUTOSTARTING_CARTRIDGE.html

All working on original hardware.
Just for your information. @GideonZ

**Comment by 0wwafa** [comment]
It will mean that the U2+ will be the only device non fully supporting one of the most common cartridge formats. No problem for me. As usual I just tried to be useful.

**Comment by GideonZ** [comment]
> Why do you create mirrors by copying memory when you can use the max_bank when you get the block in `read_chip_packet` ?
> It would be very easy just to set "max_bank" which at the moment is instead useless because it's always 255.

As I said, it wouldn't fix it. But apparently you prefer your own interpretation above that of the one who wrote the code.

**Comment by radius75** [comment]
Did I understand correctly?
Does mirroring in this case consist in
really staying on the same ROM banks despite the fact that the address points to higher ones.

There are no physical connections to switch the chip to higher banks.

You set the address but nothing happens. You still read the same data.

Ultimate does not emulate "no connections" in addressing. If it gets an address to switch to higher banks, it will switch to empty ones.

Therefore, the easiest way to double the size of ROM memory would be to copy data from lower banks to them. Then the so-called "mirroring" will be preserved.

Am I wrong?

**Comment by markusC64** [comment]
So it is. And the major problem is that the CRT file does not contain information which address lines to conntect and which to ignore.
So any attempt to emulate mirroring is guesswork.
Since you can use a 128k ROM when you're short of 64k ROMs, writing the same CRT file to physical ROM can also lead to different results regarding mirroring

What might work is to use the hardware sub type to select the mirroring size. And when loading the CRT internally duplicating. That would be well defined and no guesswork.

And no, having the FPGA doing the mirroring is a very bad idea. It costs FPGA space and you gain nothing since you can do the mirroring also at CRT loading time.

**Comment by 0wwafa** [comment]
Moreover: no matter the size of the eeprom 64/128/256k if I set a bank that does not exist, the high bits of the address will be ignored and so it will point the lower bank...

block #$20   will be block #0 on a 256K eeprom  and block number #$40 will be block 0 again on a 512K eeprom.

In other words: datablock = block_num & (totalblocks-1);

And this is how it works on real hardware (no matter the eeprom size) and on emulators.
I wonder why you need to memcopy ram. Isn't there a routine to "get a chunk" from the eeprom?


> Ultimate does not emulate "no connections" in addressing. If it gets an address to switch to higher banks, it will switch to empty ones.

precisely.


> What might work is to use the hardware sub type to select the mirroring size. And when loading the CRT internally duplicating. That would be well defined and no guesswork.

not sure of that. but from what I see in the code you can get the size when you read the CRT file. At that point:
(size/0x2000) - 1 = "DATA LINES"  and then it will be needed just an AND with the bank# poked in DE00.
DE00 is write only so there is no risk of doing anything wrong there.

**Comment by radius75** [comment]
> So any attempt to emulate mirroring is guesswork.
> Since you can use a 128k ROM when you're short of 64k ROMs, writing the same CRT file to physical ROM can also lead to different results regarding mirroring

I thought about that too. Then we manually cut half of the .bin in the hex editor. 
We double what remains. And there is a chance that it will start working. 
This is an extreme case when "mirroring" emulation will not be effective. 
Thanks for the explanation.

**Comment by GideonZ** [comment]
> Moreover: no matter the size of the eeprom 64/128/256k if I set a bank that does not exist, the high bits of the address will be ignored and so it will point the lower bank...
> 
> block #$20 will be block #0 on a 256K eeprom and block number #$40 will be block 0 again on a 512K eeprom.
> 
> In other words: datablock = block_num & (totalblocks-1);
> 
> And this is how it works on real hardware (no matter the eeprom size) 

False. The eeprom size does matter, as explained various times already. If the eeprom size is larger than the CRT rom, the other parts of the eeprom will be empty and they will be properly addressed, so you'll read an empty area in the eeprom.

>and on emulators. I wonder why you need to memcopy ram. Isn't there a routine to "get a chunk" from the eeprom?

Magic Desk is actual an Ocean 8K cart, variant 1, which means that you can turn it off by writing bit 7 to '1'. Bit 6 is unused, and bits 5-0 are the bank number.

Hence there are 64 banks of 8K, or 512K maximum size. When you load a 64K rom, only banks 0-7 are valid. When you load a 128K rom, only banks 0-15 are valid, etc. So when you program bank 8 on a 64K rom, you will read empty space and not bank 0. Unless, the rom is copied with memcpy to offsets 64K, 128K, 192K, 256K, 320K, 384K and 448K. 

> Or, even simpler: when a STA $DEXX happens: bank=bank AND maxbanks-1.

Maxbanks is not available in the hardware at this point.

> No need to copy any memory.

So without changing the FPGA logic, only memcpy will fix it.
 
> A note: From what I understand from the U2 code, `$DE00` is hardcoded. On a real MagicDesk it's not `$DE00` but `$DEXX` so putting the bank number in DE01 of DEFF achieves the same result than putting in DE00.

Wrong again. Apparently you haven't looked at or understood the code at all. The U2 takes any write in the DExx region for Ocean, thus also for MagicDesk.

It would be highly appreciated if you:

1) wouldn't make false claims without actually knowing for sure
2) accept other people's explanation

**Comment by mrdudz** [comment]
> Maxbanks is not available in the hardware at this point.

And just to mention it: Also the .crt file does NOT give this info. Or well it does - but different:

- First of all the maximum number of banks is defined by the hardware id, ie the CRT ID. For a regular "Magic Desk" this is, like Gideon said, 512kb - or if we ignore "modern" variants, it is 128kb). Both VICE and U2 (if i understood correctly) use 512kb here.
- Now to determine "max banks", all you can do is count the banks that exist in the .crt file. Or rather, check what the highest used bank number is. BUT - It is perfectly valid for a .crt file not to contain all banks, for example this is common in eg Easyflash .crt files, where "missing" banks will result in a block of 0xFF.

As a result you can not tell if the .crt file refers to (just) a 64k EPROM - or if the image goes to a 512k EPROM that just contains 0xFF in all "upper" blocks. Both interpretations are fine - and if in doubt, the CRT ID will tell you how to deal with this ambiguity (In this case: it is 512k)

The more i think about it, the more i think that what VICE currently does might actually be wrong :) But as said before, i have never seen a real "Magic Desk" cartridge and i don't know how that would behave. Chances are that infact both variants exist in the wild.

**Comment by 0wwafa** [comment]
![image](https://github.com/user-attachments/assets/2f05704b-c62f-493c-a12d-81deb31af91b)
![image](https://github.com/user-attachments/assets/39ae0119-032d-45b6-8137-922913894049)
It's quite obvious from the schematic:

DExx = DE00

and address in binary XXXXYYYY = 0000YYYY

**Comment by 0wwafa** [comment]
Just FYI:
![image](https://github.com/user-attachments/assets/5c11a364-5bf3-4fb5-b83f-f37e8eec4875)
![image](https://github.com/user-attachments/assets/16c06707-1a74-4625-bf3e-87a76e30a8fc)

Copilot gives a very good explaination, better than I could do, with a single error: there  is no "undefined behaviour" the number simply wraps around.

I hpe this clarify the situation.

**Comment by 0wwafa** [comment]
> > Maxbanks is not available in the hardware at this point.
> 
> And just to mention it: Also the .crt file does NOT give this info. Or well it does - but different:
> 
> * First of all the maximum number of banks is defined by the hardware id, ie the CRT ID. For a regular "Magic Desk" this is, like Gideon said, 512kb - or if we ignore "modern" variants, it is 128kb). Both VICE and U2 (if i understood correctly) use 512kb here.
> * Now to determine "max banks", all you can do is count the banks that exist in the .crt file. Or rather, check what the highest used bank number is. BUT - It is perfectly valid for a .crt file not to contain all banks, for example this is common in eg Easyflash .crt files, where "missing" banks will result in a block of 0xFF.
> 
> As a result you can not tell if the .crt file refers to (just) a 64k EPROM - or if the image goes to a 512k EPROM that just contains 0xFF in all "upper" blocks. Both interpretations are fine - and if in doubt, the CRT ID will tell you how to deal with this ambiguity (In this case: it is 512k)
> 
> The more i think about it, the more i think that what VICE currently does might actually be wrong :) But as said before, i have never seen a real "Magic Desk" cartridge and i don't know how that would behave. Chances are that infact both variants exist in the wild.

You can just determine that from the file size.
Since the eeproms have 8KB blocks, FILE SIZE (-header size) / 8192 = number of blocks.

**Comment by 0wwafa** [comment]
> Wrong again. Apparently you haven't looked at or understood the code at all. The U2 takes any write in the DExx region for Ocean, thus also for MagicDesk.
> 
> It would be highly appreciated if you:
> 
> 1. wouldn't make false claims without actually knowing for sure
> 2. accept other people's explanation

I didn't check U2 code (is not my job), but I made a cartridge that instead of addressing  DE00 addresses DE5A and it does not work on the U2, while the one addressing DE00 works.
The two cartridges have only one byte different.

**Comment by 0wwafa** [comment]
> Let's take a closer look at the schematic and the connections to the 74LS273 and 74LS02 to determine their exact roles.
> 
> ### 74LS273 (8-bit Latch)
> - **Data Lines (D0-D7)**: These lines carry the value to be latched when a number is stored at address DExx.
> - **Clock Input (CLK)**: This input is triggered when a write operation occurs at address DExx, causing the 74LS273 to latch the value on the data lines.
> - **Outputs (Q0-Q7)**: The latched value is output on these lines. Only the lower 3 bits (Q0-Q2) are used for bank switching.
> 
> ### 74LS02 (Quad 2-input NOR Gate)
> - **Address Decoding**: The 74LS02 is used to decode the address DExx. It combines various address lines and control signals to generate the necessary enable signals.
> - **Control Signals**: The NOR gates generate control signals that determine when the 74LS273 should latch a new value and when the memory chip should be accessed.
> 
> ### Bank Switching and Device Control
> - **Storing $80 at DExx**: When you store $80 (binary 10000000) at address DExx, the 74LS273 latches this value. The higher bit (Q7) is used to disable the device. The lower 3 bits (Q0-Q2) are ignored in this case.
> - **Storing a Positive Number at DExx**: When you store a positive number (0-7) at address DExx, the 74LS273 latches this value. The lower 3 bits (Q0-Q2) are used to select the appropriate 8KB memory bank, and the device is enabled.
> 
> ### Summary
> - The **74LS273** latches the value stored at address DExx. The lower 3 bits (Q0-Q2) are used for bank switching, and the higher bit (Q7) is used to enable or disable the device.
> - The **74LS02** provides the necessary address decoding and control signals to ensure the correct operation of the 74LS273 and the memory chip.
> 
> If you have any more questions or need further details, feel free to ask!

**Comment by 0wwafa** [comment]
Obviously the same goes with this https://github.com/msolajic/c64-uni-cart
And the same goes for ocean cartridges: https://luigidifraia.wordpress.com/2019/09/21/oceans-navy-seals-cartridge-schematics/

**Comment by 0wwafa** [comment]
![image](https://github.com/user-attachments/assets/bfeadee6-1694-446f-9ce5-d2dc27475812)

**Comment by 0wwafa** [comment]
To conclude this bug report I attach here a file with 3 cartridges.
If flashed on a real MagicDesk cartridge, they will all work.
In VICE (check its sources because they are well commented) they all work.
The "bug triggering" ones don't work on U2 and don't work on kung-fu-flash and even in MISTER, because they all made the same (wrong) assumptions.
[cartridge_bugs.zip](https://github.com/user-attachments/files/18079680/cartridge_bugs.zip)

**Comment by radius75** [comment]
> cartridge_bugs.zip)

All three versions, when doubled to 128KB, run on U2+

**Comment by GideonZ** [comment]
> As a result you can not tell if the .crt file refers to (just) a 64k EPROM - or if the image goes to a 512k EPROM that just contains 0xFF in all "upper" blocks. Both interpretations are fine - and if in doubt, the CRT ID will tell you how to deal with this ambiguity (In this case: it is 512k)

I think that it is actually possible to differentiate between 64K EPROM and 128K EPROM with half of the chip filled with 0xFF -  without changing the definition of the .crt file at all. In general, mirroring should be applied. If you want to specify a 128K EPROM with half of the space free, you can simply specify a single chunk with offset 0x10000, with 0xFFs in there. So yes, the file would be slightly larger than the 64K variant, but at least no additional variants, subtypes etc need to be defined. Similarly, if you want to specify a 512K ROM with just 64K of valid data, just add one empty chunk with the offset 0x40000 (or anything up to 0x7C000).

The loading / mirroring rule is simple:
```
fill entire emulation memory with 0xFF
load all chunks
round up the highest address to the next power of 2.
 // e.g. if loaded up to 0x10000, the rounded value would be 0x10000
 // if loaded up to 0x10001, the rounded value would be 0x20000)

while(rounded_address < max_supported_size) {
      copy from zero to rounded address -1 to rounded address
      rounded_address <<= 1
```

**Comment by GideonZ** [comment]
Message 1: 
> From what I understand from the U2 code, $DE00 is hardcoded. 

Message 2:
> I didn't check U2 code (is not my job), but [...]

You are contradicting yourself, @0wwafa .  

If you are experiencing hostility here, please consider the following:

Posting all these schematics and references makes absolutely zero sense if it shows that we already understand what the issue is. The problem here is that you don't accept our explanation and assume we don't understand. Read carefully what @markusC64 and @mrdudz mention, from there you will see that we understand perfectly how the hardware works and what needs to be done to allow sloppy cartridge code. Of course, the real issue is in _your_ cartridge code, where the wrong assumption is made that bank 0x08 points to anything valid (bank 0). This may be true with real hardware under certain conditions, but that doesn't make it any less wrong.

I'll leave this issue open, because the mirroring needs to be fixed. But the discussion is hereby **closed**.

**Comment by 0wwafa** [comment]
> Message 1:
> 
> > From what I understand from the U2 code, $DE00 is hardcoded.
> 
> Message 2:
> 
> > I didn't check U2 code (is not my job), but [...]
> 
> You are contradicting yourself, @0wwafa .
> 
> If you are experiencing hostility here, please consider the following:
> 
> Posting all these schematics and references makes absolutely zero sense if it shows that we already understand what the issue is. The problem here is that you don't accept our explanation and assume we don't understand. Read carefully what @markusC64 and @mrdudz mention, from there you will see that we understand perfectly how the hardware works and what needs to be done to allow sloppy cartridge code. Of course, the real issue is in _your_ cartridge code, where the wrong assumption is made that bank 0x08 points to anything valid (bank 0). This may be true with real hardware under certain conditions, but that doesn't make it any less wrong.
> 
> I'll leave this issue open, because the mirroring needs to be fixed. But the discussion is hereby **closed**.

Sorry if I misunderstood.
Anyway, it's not only the "mirroring", Also DExx vs DE00. Don't forget.

**Comment by 0wwafa** [comment]
By the way, https://github.com/GideonZ/1541ultimate/issues/442 is unrelated to this and does not involve "sloppy" code (code that works well is never sloppy!)
What *indeed* are sloppy are emulators. And by fixing these problems they will all be better and less "sloppy".

**Comment by GideonZ** [comment]
> Anyway, it's not only the "mirroring", Also DExx vs DE00. Don't forget.

```
                if io_write='1' and io_addr(8)='0' then -- DE00 range
```
You can test again; the whole DExx region works; as the remaining address pins are not even checked; only A(8) as you can see.

**Comment by GideonZ** [comment]
> [..] code that works well is never sloppy! [..]

You're a hopeless case.

**Comment by 0wwafa** [comment]
> > [..] code that works well is never sloppy! [..]
> 
> You're a hopeless case.

You see? In all this I exposed a BUG in *YOUR* software, and the hopless case is me?
That's what I meant when I said that I feel hostility.
I point out a bug. And you take it on me.
Meh.
